### PR TITLE
[tests-only] fix space ID calculation for a non existent user

### LINKED ISSUE #4820: Cannot test trashbin of non existing user via spaces
## Describe the bug

The testsuite cannot look up the space of a non existing user:
```
@issue-ocis-3561 @skipOnLDAP @skip_on_objectstore
Scenario Outline: Listing non-existent user's trashbin is prohibited                   # /drone/src/tmp/testrunner/tests/acceptance/features/apiTrashbin/trashbinFilesFolders.feature:277
  Given using <dav-path> DAV path                                                      # FeatureContext::usingOldOrNewDavPath()
  When user "Alice" tries to list the trashbin content for user "testtrashbinnotauser" # TrashbinContext::userTriesToListTheTrashbinContentForUser()
  Then the HTTP status code should be "404"                                            # FeatureContext::thenTheHTTPStatusCodeShouldBe()

  Examples:
    | dav-path |
    | new      |
    | spaces   |
      Failed step: When user "Alice" tries to list the trashbin content for user "testtrashbinnotauser"
      TestHelpers\WebDavHelper::getPersonalSpaceIdForUser webdav propfind for user testtrashbinnotauser failed with status 401 - so the personal space id cannot be discovered (Exception)
```

Which all makes sense, because it is impossible to look up the space for a user.

AFAICT the only way to test this is by using a non existing spaceid and sending a propfind to its (non existing) trashbin endpoint.

### PR DESCRIPTION
**Pull request by kiranparajuli589**
## Description
With this PR, for the `makeDavRequest` function, if the personal space id getter function throws exception, then a fake space id is assigned as the user's space id.

## Related Issue

- Fixes https://github.com/owncloud/ocis/issues/4820

## How Has This Been Tested?

- https://github.com/owncloud/ocis/pull/4840

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUISharingIntGroup1-chrome-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/37033/137

File: tests/TestHelpers/WebDavHelper.php [hunk 1]
```diff
@@ -179,8 +179,10 @@ public static function getBodyForPropfind(?array $properties): string {
 	 * @param string|null $type
 	 * @param int|null $davPathVersionToUse
 	 * @param string|null $doDavRequestAsUser
+	 * @param bool|null $nonExistentUser
```
Comment (saw-jan) [inline tests/TestHelpers/WebDavHelper.php:182]: the default is `false` so I think, better to not allow null
```suggestion
	 * @param bool $nonExistentUser
```

File: tests/TestHelpers/WebDavHelper.php [hunk 2]
```diff
@@ -545,6 +549,7 @@ public static function getPersonalSpaceIdForUser(string $baseUrl, string $user,
 	 * @param Client|null $client
 	 * @param array|null $urlParameter to concatenate with path
 	 * @param string|null $doDavRequestAsUser run the DAV as this user, if null its same as $user
+	 * @param bool|null $forNonExistingUser run the DAV as a non-existing user
```
Comment (saw-jan) [inline tests/TestHelpers/WebDavHelper.php:552]: the default is false so I think, better to not allow null
```suggestion
	 * @param bool $forNonExistingUser run the DAV as a non-existing user
```

File: tests/acceptance/features/bootstrap/TrashbinContext.php [hunk 3]
```diff
@@ -374,11 +375,12 @@ public function theTrashbinDavResponseShouldContainTheseNodes(TableNode $table):
 	 * @param string $user user
 	 * @param string|null $asUser - To send request as another user
 	 * @param string|null $password
+	 * @param bool|null $nonExistentUser (if trying to request on behalf of a non-existent user)
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/TrashbinContext.php:378]: the default is false so I think, better to not allow null
```suggestion
	 * @param bool $nonExistentUser (if trying to request on behalf of a non-existent user)
```

File: tests/TestHelpers/WebDavHelper.php (hunk 2 above)
Comment (kiranparajuli589) [reply tests/TestHelpers/WebDavHelper.php]: true. fixed!

**Review by saw-jan — COMMENTED** [review]
LGTM

**Comment by individual-it** [comment]
I can see two options that would avoid an other parameter passing through all layers
1. in `getPersonalSpaceIdForUser()` check if the user exists, or was ever created during the tests. If not, do not try to find a space-id but make one up
2. catch the exception `getPersonalSpaceIdForUser` throws when it cannot find a space id and then fake a space id

**Comment by kiranparajuli589** [comment]
> 1. if the user exists, or was ever created during

Yes, the list of the params. for `makeDavRequest` is huge already. I've tried to implement the first idea you suggested. please review.

File: tests/TestHelpers/WebDavHelper.php [hunk 4]
```diff
@@ -419,17 +419,32 @@ public static function listFolder(
...
-	public static function getPersonalSpaceIdForUser(string $baseUrl, string $user, string $password, string $xRequestId):string {
-		if (\array_key_exists($user, self::$spacesIdRef) && \array_key_exists("personal", self::$spacesIdRef[$user])) {
+	public static function getPersonalSpaceIdForUser(
+		string $baseUrl,
+		string $user,
+		string $password,
+		string $xRequestId,
+		bool $returnFakeIfNotFound = false
+	):string {
+		$fakeId = "fake-personal-space-id";
```
Comment (individual-it) [inline tests/TestHelpers/WebDavHelper.php:443]: what about making it a UUID?

File: tests/TestHelpers/WebDavHelper.php [hunk 5]
```diff
@@ -458,6 +473,7 @@ public static function getPersonalSpaceIdForUser(string $baseUrl, string $user,
 			// we expect to get a multipart XML response with status 207
 			$status = $response->getStatusCode();
 			if ($status !== 207) {
+				if ($returnFakeIfNotFound) return $fakeId;
```
Comment (individual-it) [inline tests/TestHelpers/WebDavHelper.php:476]: This is dangerous. A not careful reformatting could kick the return out of the if-statement 
```suggestion
				if ($returnFakeIfNotFound) {
 				 return $fakeId;
				 }
```

**Review by individual-it — COMMENTED** [review]
I'm generally happy with the result, but looking at the code changes a try-catch block in `makeDavRequest()` might be even simpler.
It's up to you

**Review by individual-it — APPROVED** [review]

File: tests/TestHelpers/WebDavHelper.php [hunk 6]
```diff
@@ -416,6 +416,23 @@ public static function listFolder(
...
+		// generate 16 bytes (128 bits) of random data or use the data passed into the function.
+		$data = random_bytes(16);
+		assert(strlen($data) == 16);
+
+		$data[6] = chr(ord($data[6]) & 0x0f | 0x40); // set version to 0100
+		$data[8] = chr(ord($data[8]) & 0x3f | 0x80); // set bits 6-7 to 10
+
+		return vsprintf('%s%s-%s-%s-%s-%s%s%s', str_split(bin2hex($data), 4));
+	}
+
```
Comment (saw-jan) [inline tests/TestHelpers/WebDavHelper.php:435]: There is already one but I think that cannot be used here (`Context` to `Helper`)
https://github.com/owncloud/core/blob/ffc3c40b5b0a7693818dc03b485f12794920fad4/tests/acceptance/features/bootstrap/Provisioning.php#L762-L767

Comment (individual-it) [reply tests/TestHelpers/WebDavHelper.php]: then we should change all the usages to the helper

Comment (kiranparajuli589) [reply tests/TestHelpers/WebDavHelper.php]: hmm..thx for finding ths. I'll try to remove this duplication.

**Review by individual-it — APPROVED** [review]

File: tests/TestHelpers/WebDavHelper.php [hunk 7]
```diff
@@ -578,20 +597,25 @@ public static function makeDavRequest(
...
-				$spaceId = self::getPersonalSpaceIdForUser($baseUrl, $doDavRequestAsUser, $password, $xRequestId);
+			try {
+				$spaceId = self::getPersonalSpaceIdForUser(
+					$baseUrl,
+					$doDavRequestAsUser ?? $user,
+					$password,
+					$xRequestId,
+				);
+			} catch (Exception $e) {
+				// if the fetch fails, and the user is not found, then the a fake space id is prepared
```
Comment (phil-davis) [inline tests/TestHelpers/WebDavHelper.php:608]:
```suggestion
				// if the fetch fails, and the user is not found, then a fake space id is prepared
```

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/core/37066/27/10
```
There were 85 errors:

1) OCA\Files_External\Tests\Storage\OwncloudTest::testRoot
OCP\Files\StorageNotAvailableException: Sabre\HTTP\ClientHttpException: Bad Request
```
That is a strange error in a PHP unit test pipeline. I restarted CI.

**Review by phil-davis — APPROVED** [review]

**Comment by phil-davis** [comment]
> That is a strange error in a PHP unit test pipeline. I restarted CI.

Issue #40436 which will be fixed by PR #40438 - after that is merged, then this can be rebased...

**Comment by phil-davis** [comment]
Rebased just now - this can be merged when CI is green.

**Pull request merged**